opentelemetry-api/src/opentelemetry/trace/span.py

```diff
@@ -118,6 +118,23 @@ def add_event(
         timestamp if the `timestamp` argument is omitted.
         """
 
+    @abc.abstractmethod
```

Comment:
I remember there being a lot of discussion in the spec about how to make adding to the API backward compatible. Existing implementations of `Span` would break with this new API change as they wouldn't be implementing all the abstract methods.

We should probably include a default noop implementation and not mark this as abstract. The default implementation should never be called normally (as real `Span` implementations should implement it), and we could add a warning letting the user know that `Span.add_link()` is doing nothing.